Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,306 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.ObjectBinding.Validation;
+using Remotion.Utilities;
+
+namespace Remotion.ObjectBinding.Web.UI.Controls.Validation
+{
+  public class BocListValidationFailureRepository
+  {
+    private class HandleableFailure
+    {
+      public BusinessObjectValidationFailure Failure { get; }
+
+      public bool Handled { get; private set; }
+
+      public HandleableFailure (BusinessObjectValidationFailure failure)
+      {
+        Failure = failure;
+        Handled = false;
+      }
+
+      public void MarkAsHandled ()
+      {
+        Handled = true;
+      }
+    }
+
+    private class RowFailureContainer
+    {
+      private List<HandleableFailure>? _rowFailures;
+      private Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>? _cellFailures;
+
+      public void AddRowFailures (IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        _rowFailures ??= new List<HandleableFailure>();
+        _rowFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void AddCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition, IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        var cellFailures = GetOrCreateListOfCellFailures(cellDefinition);
+        cellFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void GetAllFailures (ICollection<BusinessObjectValidationFailure> failures, bool markAsHandled)
+      {
+        GetRowFailures(failures, markAsHandled);
+        GetAllCellFailuresForRow(failures, markAsHandled);
+      }
+
+      public void GetRowFailures (ICollection<BusinessObjectValidationFailure> failures, bool markAsHandled)
```

Comment:
Querstion: are these mthods used in multiple places with markashandled semantics or just once. it's possible we don't need mark as handled in the nested type but can keep that in the parent class. this would make the nested class a stupid collecton struture without business logic. might be easier to maintain.